### PR TITLE
Get draft content item by alias

### PR DESCRIPTION
**Pull request by deanmarcussen**
Fixes https://github.com/OrchardCMS/OrchardCore/issues/5186

Looked like the best option was to add the published column as per autoroute.

Also allows now for getting an autoroute alias `{slug:path}` that is a draft only (or changes between draft and published)

I still need to test this properly

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Alias/Indexes/AliasPartIndex.cs [hunk 1]
```diff
@@ -1,13 +1,15 @@
...
 using YesSql.Indexes;
+using System;
 
 namespace OrchardCore.Alias.Indexes
 {
     public class AliasPartIndex : MapIndex
     {
         public string ContentItemId { get; set; }
         public string Alias { get; set; }
+        public bool Published { get; set; }
```
Comment (carlwoodhouse) [inline src/OrchardCore.Modules/OrchardCore.Alias/Indexes/AliasPartIndex.cs]: you add this to the index but don't actually set it; or use it as far as i can see ?

File: src/OrchardCore.Modules/OrchardCore.Autoroute/Services/AutorouteAliasProvider.cs [hunk 2]
```diff
@@ -26,14 +30,18 @@ public Task<string> GetContentItemIdAsync(string alias)
...
                 }
 
-                string contentItemId;
-                if (_autorouteEntries.TryGetContentItemId(alias, out contentItemId))
+                // This only contains published items.
+                if (_autorouteEntries.TryGetContentItemId(alias, out var contentItemId))
                 {
-                    return Task.FromResult(contentItemId);
+                    return contentItemId;
+                } else // Not relevant to check against published here as we are only querying for the content item id for the path / alias provided.
```
Comment (carlwoodhouse) [inline src/OrchardCore.Modules/OrchardCore.Autoroute/Services/AutorouteAliasProvider.cs]: this could potentially cause lots of db lookups; as its going too query even if someone attempts to lookup an alias that doesn't exist ... whereas before; if it didnt exist this would just return null and let another IContentAliasProvider potentially pick it up :|

File: src/OrchardCore.Modules/OrchardCore.Alias/Indexes/AliasPartIndex.cs (hunk 1 above)
Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.Alias/Indexes/AliasPartIndex.cs]: Also why not Latest AND Published fields. It won't change the number of records, but you can then also get explicitly the latest version whatever it is. Though there can be only one record in latest, and one record in published (they might both be true too, so one record).

File: src/OrchardCore.Modules/OrchardCore.Alias/Migrations.cs [hunk 3]
```diff
@@ -27,13 +27,25 @@ public int Create()
...
-            return 1;
+            // Return 2 to shortcut the second migration on new content definition schemas.
+            return 2;
+        }
+
+        // This code can be removed in a later version as published is a recent addition.
+        public int UpdateFrom1()
+        {
+            SchemaBuilder.AlterTable(nameof(AliasPartIndex), table => table
+                .AddColumn<bool>("Published")
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Alias/Migrations.cs]: Maybe have `true` as a default value so that existing setups will work.

File: src/OrchardCore.Modules/OrchardCore.Autoroute/Services/AutorouteAliasProvider.cs (hunk 2 above)
Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.Autoroute/Services/AutorouteAliasProvider.cs]: I would not change this service.

File: src/OrchardCore.Modules/OrchardCore.Alias/Migrations.cs (hunk 3 above)
Comment (deanmarcussen) [reply src/OrchardCore.Modules/OrchardCore.Alias/Migrations.cs]: Done

File: src/OrchardCore.Modules/OrchardCore.Autoroute/Services/AutorouteAliasProvider.cs (hunk 2 above)
Comment (deanmarcussen) [reply src/OrchardCore.Modules/OrchardCore.Autoroute/Services/AutorouteAliasProvider.cs]: Reverted.
Noting that it wouldn't cause extra lookups as the `AutorouteAliasProvider` checks for `.StartsWith("slug")` before it performs any lookups.
So without this you can get an alias draft, but not an autoroute draft (it will have to be published to be in the `AutorouteEntries`)

File: src/OrchardCore.Modules/OrchardCore.Alias/Indexes/AliasPartIndex.cs (hunk 1 above)
Comment (deanmarcussen) [reply src/OrchardCore.Modules/OrchardCore.Alias/Indexes/AliasPartIndex.cs]: No, totally forgot to set it ;)

Added Latest as well

**Comment by deanmarcussen** [comment]
On reflection this does not need to have either `Published` or `Latest` columns

For the purposes of `.GetContentItemByAlias` this will now track both published and draft items.
If the alias was changed between the two (the edge case) it will still return the `ContentItemId` when the correct alias is searched for, and even if the `Published` column was tracked it would still require the correct alias.

The method only returns a `ContentItemId`, which will be the same for both `Published` or `Latest`, and it then makes another query based on the `ContentItemIndex` to decide which content item version to return.

So there is no need to track the `Published` or `Latest` status in the `AliasPartIndex` as it is never used.

But in the edge case where the alias changes between published and draft it will still create two records in the index table, with the different alias **but the same `ContentItemId`** and then the `ContentItem` is published it will remove the second record.

Moving away from the context of decoupled where this is the issue even then a query against the index would be much more like

```
session.Query<ContentItem>()
    .With<AliasPartIndex>(x => x.Alias == "myalias")
    .With<ContentItemIndex>(x => x.Published == true);    
```

and unless we wanted to start tracking the `ContentItemVersionId` there is no benefit that I can see from having the version records in the `AliasPartIndex` when the way to access them is a cross ref against the `ContentItemIndex`

For the purposes of the `AutoroutePartIndex` it has a good use case there, because in that case we want to be able to get all the correct information for building routes from the index table, rather than having to join on the `Document` table as well... i.e.
`var autoroutes = session.QueryIndex<AutoroutePartIndex>(o => o.Published).ListAsync();`

But I can't see a use case for doing this with the `AliasPartIndex` as any query would always want to return the `ContentItem` as well as the query predicate against the index.

**Comment by sebastienros** [comment]
You do need Latest and Published, so you can support when the alias is changed in a Draft, and you want to preview the draft while still serving the Published one publicly. And you also need to filter with `Published || Latest` in the index such that it gets rid of previous versions.

**Comment by sebastienros** [comment]
I understand your comment, but like any "index", not having to do a join with another table is still valuable for performance.

**Comment by deanmarcussen** [comment]
> I understand your comment, but like any "index", not having to do a join with another table is still valuable for performance.

Cool, just pointing out that it wasn't necessary to fix this issue so have reverted it.

It is filtering on `Published || Latest` (I think) so one record created for both entries, when both are published and latest, multiple records created when they are different (and for alias changing), and then those records removed when publishing. 

Noting : for performance indexes like this should probably also index the published and latest properties (in the column indexes) - to get the best sql execution plan, but given there are a few issues with index lengths will leave for the moment, and perhaps review some indexes (autoroute in particular) and sql execution plans, if / when get to YesSql

File: src/OrchardCore.Modules/OrchardCore.Alias/Migrations.cs [hunk 4]
```diff
@@ -27,13 +27,30 @@ public int Create()
...
-            return 1;
+            // Return 2 to shortcut the second migration on new content definition schemas.
+            return 2;
+        }
+
+        // This code can be removed in a later version as Latest and Published are alterations.
+        public int UpdateFrom1()
+        {
+            SchemaBuilder.AlterTable(nameof(AliasPartIndex), table => table
+                .AddColumn<bool>("Latest", c => c.WithDefault(true))
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Alias/Migrations.cs]:
```suggestion
                .AddColumn<bool>("Latest", c => c.WithDefault(false))
```

Because it's just for existing records that will be empty, and these are `false`

File: src/OrchardCore.Modules/OrchardCore.Alias/Migrations.cs [hunk 5]
```diff
@@ -27,13 +27,30 @@ public int Create()
             SchemaBuilder.CreateMapIndexTable(nameof(AliasPartIndex), table => table
                 .Column<string>("Alias", col => col.WithLength(AliasPartDisplayDriver.MaxAliasLength))
                 .Column<string>("ContentItemId", c => c.WithLength(26))
+                .Column<bool>("Latest", c => c.WithDefault(true))
+                .Column<bool>("Published", c => c.WithDefault(true))
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Alias/Migrations.cs]: I would index these columns as most queries will use it. And it's safe.

**Comment by sebastienros** [comment]
I am not sure this is sufficient to get the drafts previews working with alias. I can test it if you prefer.

**Comment by deanmarcussen** [comment]
Putting this on hold for a few days, while I am looking at the index key length issue that exists with Alias for both MS SQL and MySql.

May have to reduce the length. Maybe other options. Halfway through figuring out the best ones

But will test against drafts preview (it's in the docs right?) when I have figured the other issue.

**Comment by sebastienros** [comment]
It's in the Decoupled CMS guide. With the PreviewPart. Test with a content item that has never been published.

**Comment by deanmarcussen** [comment]
Tested
- Content Item never published works with preview
- Content Item with different alias in draft works
- Public route still works with the published alias.

But the new alias, `sample-alias-draft` also works as a published route alongside the `sample-alias-orig`, although it serves the old content.

Have a good set of failing tests for YesSql showing the key length issues. so WIP

File: src/OrchardCore.Modules/OrchardCore.Alias/Migrations.cs (hunk 4 above)
Comment (deanmarcussen) [reply src/OrchardCore.Modules/OrchardCore.Alias/Migrations.cs]: Done

File: src/OrchardCore.Modules/OrchardCore.Alias/Migrations.cs (hunk 5 above)
Comment (deanmarcussen) [reply src/OrchardCore.Modules/OrchardCore.Alias/Migrations.cs]: Done.

Have adjusted alias maximum size to 767, based on tests against MySql and MS SQL index key length maximums.
767 under utf8 for MySql is the largest the Alias can go based on 4 char * 767 = 3068 + 2 for the `Latest` and `Published` index keys = `3070` **the maximum key length for a MySql index.**

This is not far off the maximum for MS SQL under default collation of 1700 bytes (uses less per charactor)

Have also placed better testing against maximum lengths in the `AliasHandler`, and tested.

Will apply some of this code to the `AutorouteHandler` (without the changes to length as we do not index, not need to index the `Path` property) but in another PR

**Comment by deanmarcussen** [comment]
See above.

Tested and preview works with draft content items in decoupled.

I cannot find a way to stop the edge case of the alias changing and working as a published route while still in draft. Any checks against `Latest` stop the preview working.

Note: It serves under the new alias route, but serves the old content under the new alias. So not a big issue I think as there is no way except with preview mode to see the drafted content.

**Comment by jtkech** [comment]
> I cannot find a way to stop the edge case of the alias changing and working as a published route while still in draft. Any checks against Latest stop the preview working.

Can you elaborate a little to see if i could help on this edge case? Do you know that when previewing we work on a fake content item whose `Id = -1`?

**Comment by sebastienros** [comment]
We can follow up on the comment from @jtkech later.

**Comment by deanmarcussen** [comment]
> Can you elaborate a little to see if i could help on this edge case? Do you know that when previewing we work on a fake content item whose `Id = -1`?

@jtkech I'm not sure if there is a way around this (but I do think its minor). Let me explain the sequence that's happening.

So the `PreviewPart` gets the url it should point the preview page at from the `Pattern` on the part.

That page calls `.GetContentItemByAliasAsync("slug:the-draft-alias")`

`GetContentItemByAliasAsync` searches the index for the alias, **not caring if it is a draft or published** it just returns the `ContentItemId` 
It then calls `contentManager.GetAsync(contentItemId)` with the options for latest or published

The preview swaps out the contentitem from the cache with the draft preview item.

But in order for the process to work the razor page must be able to call `.GetContentItemByAliasAsync("slug:the-draft-alias")`
without getting a null reference because it doesn't exist (or doesn't find it if you start sending `latest/published` into the `IContentAliasManager`)

Which means that in order to be able to preview a draft, razor pages needs to be able to route to that alias.

Does that make sense?

Regardless in non preview mode the `draft-alias` route will still only display the published content.

So there is no exposure of draft content, just the drafted alias...

**Comment by jtkech** [comment]
Okay i understand but i would need to try it, we will see if it is a real issue ;)

Not sure it could help but for infos when we create a fake content item for previewing and that we cache in the `_contentManagerSession`, we set `Published` to true but not `Latest`.
